File: JAVA/CycleLinkList.java

/*
* author: acme
*   date: 2017-1-19
*  blogs: http://blog.csdn.net/qq_18297675
*/

public class CycleLinkList {

	private Node head = new Node(0);    //头节点
	private Node tail = new Node(0);    //尾指针
	private int iCount = 0;      		//链表大小
	
	public CycleLinkList() {
		head.next = head;          //初始化的时候，自己指向自己
	}
	
	//判断表是否为空
	public boolean isEmpty() {
		return iCount == 0 ? true : false;
	}
	
	//获取元素个数
	public int size() {
		return iCount;
	}
	
	//添加数据（头添加）
	public void addFromHead(Node node) {	
		if (isEmpty()) //如果链表为空，则尾指针指向第一次添加进来的节点
			tail.next = node;
		node.next = head.next;
		head.next = node;
		++iCount;
	}
	
	//添加数据（尾添加）
	public void addFromTail(Node node) {		
		Node cur = head;
		for (int i = 0; i < iCount; i++) {
			cur = cur.next;          //遍历到链表末尾
		}
		cur.next = node;               //直接把新的节点挂在链表末尾
		node.next = head;            //最后一个节点指向head
		tail.next = node;            //tail指向最后一个节点
		++iCount;
	}
	
	//插入数据（指定位置）
	public void insert(int local, Node node) {	
		Node cur = head;   //注意，这里千万不能是m_head->next,因为如果当前链表为空插入的位置
		 				   //是0，则会直接执行cur->next这个表达式，解引用空指针会出错							
		if (local < iCount && local >= 0) {
			for (int i = 0;i < local;i++, cur = cur.next); //遍历到local这个位置
			node.next = cur.next;
			cur.next = node;       //这里的操作就和头添加一样,就是改变指向
			++iCount;
		}
		else if(local == iCount)//如果是插入尾部，直接调用AddFromTail函数即可
			addFromTail(node);
	}
	
	//删除数据 （指定位置）
	public void deleteFromLocal(int local) {		
		if (local < iCount && local >= 0) {//当删除的位置合法的时候才进行删除操作
			Node pre = head;
			Node cur = head.next;
			for (int i = 0;i < local;i++, cur = cur.next, pre = pre.next);  //遍历到指定位置
			pre.next = cur.next;  //例如:pre->cur->next,现在直接让pre->next
			--iCount;
		}
	}
	
	//删除数据 （指定元素）
	public void deleteFromElement(int e) {			
		Node pre = head;
		Node cur = head.next;
		for (int i = 0;i < iCount;i++, cur = cur.next, pre = pre.next)  //遍历到指定位置
			if (cur.data == e) {
				pre.next = cur.next;  //例如:pre->cur->next,现在直接让pre->next
				--iCount;
				break;
			}
	}
	
	//根据指定元素查找，返回位置
	public int searchFromElmemt(int e) {				
		return getLocal(e);
	}
	
	//根据位置查找指定元素  
	public int searchFromLocal(int local) {		
		if (local < iCount && local >= 0) {//当搜索的位置合法的时候才进行搜索操作
			Node cur = head.next;
			for (int i = 0; i < local; i++, cur = cur.next);
			return cur.data;
		}
		else
			return -1;
	}
	
	////逆转表(方法1，牺牲时间换取空间，速度很慢，时间复杂度为0(n^2))
	public void reverse1() {						
		int temp;
		for (int i = 0;i < iCount / 2;i++) {
			Node node1 = getNode(i);                   //获取节点
			Node node2 = getNode(iCount - i - 1);
			temp = node1.data;                         //交换数据项，千万别交换节点
			node1.data = node2.data;
			node2.data = temp; 
		}
	}
	
	//逆转表(方法2，牺牲空间换取时间，速度很快，时间复杂度为0(n))
	public void reverse2() {
		Node tHead = new Node(0);     //临时头节点
		Node tTail = new Node(0);     //临时尾指针
		tHead.next = tHead;
		Node cur = head.next;
		for (int i = 0;i < iCount;i++) {
			Node node = new Node(0);
			node = (Node)cur.clone();		//这里一定要复制，而不是赋值=，如果是=，则两指针指向的是同一个节点
			if(i == 0)
			{
				tTail.next = node;           //一开始就让tTail指向最后一个节点		 
				node.next = tHead;			 //最后一个节点指向头节点
			}
			node.next = tHead.next;			  //这里采用头添加，结束后，就会自动逆转过来了         
			tHead.next = node;
			cur = cur.next;
		}
		head = tHead;   //因为头添加已经是逆转了，所以直接把头节点和尾节点的地址赋值给成员变量即可
		tail = tTail;   //注意，一定要头尾同时赋值，不然它们的指向就乱了。
	}
	
	//遍历元素
	public void print() {
		if(isEmpty())   //如果链表为空，则不做任何操作，防止空指针异常
			return;
		Node cur = head.next;
		for (int i = 0; i < iCount; i++,cur = cur.next) {
			System.out.println(cur.data);
		}
	}
	
	//复制表
	public CycleLinkList copyList() {	//不能直接用clone方法，必须要一个个节点复制
		CycleLinkList list = new CycleLinkList();
		Node cur = head.next;
		for (int i = 0; i < iCount; i++,cur = cur.next) {
			Node temp = (Node)cur.clone();  //节点复制,不要直接赋值引用
			list.addFromHead(temp);  //头添加效率高
		}
		list.reverse2();        //逆转回来
		return list;
	}
	
	//修改元素
	public void rewrite(int local, int e) {			
		if (local < iCount && local >= 0) {
			Node cur = head.next;
			for (int i = 0;i < local;i++, cur = cur.next);
			cur.data = e;
		}
	}
	
	//清空链表
	public void clearLinkList() {                   
		head.next = head;       //因为java有垃圾回收机制，所以不需要手动释放节点，只要让头节点指向自己,尾指针为空就好
		tail.next = null;
		iCount = 0;
	}
	
	//获取指定位置的节点
	public Node getNode(int local) {            
		if (local < iCount && local >= 0) {//当获取的位置合法的时候才进行获取操作
			Node cur = head.next;
			for (int i = 0;i < local;i++, cur = cur.next);
			return cur;
		}
		else
			return null;
	}
	
	//获取最后一个节点
	public Node getTailNode() {                 
		return tail.next;
	}
	
	//返回元素位置
	private int getLocal(int e) { 						
		Node cur = head.next;
		for (int i = 0;i < iCount;i++, cur = cur.next)
			if (cur.data == e)
				return i;
		return -1;
	}
	
	//升序
	public void sortByAsc() {
		selectSort(true);
	}
	
	//降序
	public void sortByDesc() {
		selectSort(false);
	}
	
	//选择排序法
	private void selectSort(boolean bAsc) {
		for (int i = 0;i < iCount;i++) {
			Node node1 = getNode(i);
			for (int j = i + 1;j<iCount;j++) {
				Node node2 = getNode(j);
				int temp;
				if (bAsc) {//升序
					if (node1.data > node2.data) {
						temp = node1.data;
						node1.data = node2.data;
						node2.data = temp;
					}
				}
				else  {//降序
					if (node1.data < node2.data) {
						temp = node1.data;
						node1.data = node2.data;
						node2.data = temp;
					}
				}
			}
		}
	}
	//合并链表  ,需要注意的是，合并后，原来的表就会被改变。
	public void mergeList(CycleLinkList sl) {
		tail.next.next = sl.head.next;     //表1的最后一个节点指向表2的第一个节点
		tail.next = sl.tail.next;           //表1的尾指针指向表二的最后一个元素 
		sl.tail.next.next = head;			//表2的最后一个节点指向表1的head节点
		iCount += sl.iCount;                  //还要把元素个数加起来
	}

	//合并链表(有序)
	public void mergeListOrder(CycleLinkList sl,boolean bAsc)
	{
		mergeList(sl);
		selectSort(bAsc);
	}

	//合并链表(无序)，不会改变原表
	void mergeListNoChange(CycleLinkList sl)
	{
		CycleLinkList list = new CycleLinkList();
		list = sl.copyList();            //复制链表,这样就不会改变原表
		mergeList(list);          
	}

	//合并链表(有序),不会改变原表
	void mergeListNoChangeOrder(CycleLinkList sl, boolean bAsc)
	{
		CycleLinkList list = new CycleLinkList();
		list = sl.copyList();           //复制链表,这样就不会改变原表
		mergeListOrder(list, bAsc);
	}
}


class Node implements Cloneable{  //实现该接口后，可进行对象的复制
	
	public int data;
	public Node next;
	Node(int d) {
		data = d;
	} 
	
	//重写clone  用来进行浅复制
	public Object clone() {
		Node node = null;
		try {
			node = (Node)super.clone();
		} catch (CloneNotSupportedException e) {
			e.printStackTrace();
		}
		return node;
	}
}

File: C/CycleLinkList/CycleLinkList.c
/*
* author: acme
*   date: 2017-1-19
*  blogs: http://blog.csdn.net/qq_18297675
*/

#include <stdio.h>
#include <string.h>
#include <stdlib.h>

//学生信息结构
typedef struct _student
{
	int id;
	char name[16];
}STUDENT, *LPSTUDENT;

//链表结构
typedef struct _link_list
{
	STUDENT s;           //链表中的节点的数据项，这里用一个结构体表示
	_link_list* next;    //指向下一个节点
}CYCLE_LINK_LIST, *LPCYCLE_LINK_LIST;

CYCLE_LINK_LIST head = { 0 };   //创建一个全局的链表头节点(可以用一个函数创建，返回一个局部头节点)
                         //如果头节点是全局的，那么只能有一个链表，如果想要N个，则需要创建一个局部头节点
						 //还可以加一个尾指针，这样合并表的时候只需要把尾指针指向另一个表的头就可以了

int g_iCount = 0;        //记录当前链表中元素的个数

//初始化链表
void InitLinkList()
{
	head.next = &head;	 //其实也就是把头指向头 head->head(刚开始的时候)
}

//判断表是否为空						 
bool IsEmpty()
{
	return g_iCount ? false : true;
}

//获取元素个数
int Size()
{
	return g_iCount;
}

//添加数据（头添加）
void AddFromHead(LPCYCLE_LINK_LIST l)
{
	l->next = head.next;     //例如 head->l1,现在添加了个l,则l->l1,head->l,就是改变指向
	head.next = l;
	++g_iCount;
}

//添加数据（尾添加）
void AddFromTail(LPCYCLE_LINK_LIST l)
{
	LPCYCLE_LINK_LIST cur = &head;
	while (cur->next != &head)//循环链表，遍历到末尾不再是指针域为空了，而是等于头节点
		cur = cur->next;      //遍历到链表末尾
	cur->next = l;            //把l挂在链表的末尾
	l->next = &head;           //最后让l指向头节点
	++g_iCount;
}

//插入数据（指定位置）
void Insert(int local, LPCYCLE_LINK_LIST l)
{
	if (local < g_iCount && local >= 0)   //当要插入的位置合理的情况下才能插入
	{
		LPCYCLE_LINK_LIST cur = &head;
		for (int i = 0;i < local;i++)    //遍历到指定位置
			cur = cur->next;
		l->next = cur->next;              //在该位置插入
		cur->next = l;
		++g_iCount;
	}
	else if (local == g_iCount)
		AddFromTail(l);
}

//根据id返回元素位置
int GetLocalFromId(int id)
{
	LPCYCLE_LINK_LIST cur = head.next;
	for (int i = 0;i < g_iCount;i++)  //遍历，一一比较
	{
		if (cur->s.id == id)
			return i;
		cur = cur->next;
	}
	return -1;   //不存在，返回-1
}

//根据name返回元素位置
int GetLocalFromName(char* name)
{
	LPCYCLE_LINK_LIST cur = head.next;
	for (int i = 0;i < g_iCount;i++)  //遍历，一一比较
	{
		if (strcmp(cur->s.name, name) == 0)
			return i;
		cur = cur->next;
	}
	return -1;   //不存在，返回-1
}

//删除数据 （指定位置）
void DeleteFromLocal(int local)
{
	if (local < g_iCount && local >= 0)   //当要删除的位置合理的情况下才能删除
	{
		LPCYCLE_LINK_LIST cur = &head;
		for (int i = 0;i < local;i++)    //遍历到指定位置
			cur = cur->next;
		LPCYCLE_LINK_LIST temp;
		temp = cur->next;
		cur->next = temp->next;           //删除该位置下的元素
		--g_iCount;
		free(temp);                       //释放掉该节点
	}
}

//删除数据 （指定id）
void DeleteFromId(int id)
{
	LPCYCLE_LINK_LIST cur = head.next;
	LPCYCLE_LINK_LIST pre = &head;
	for (int i = 0;i < g_iCount;i++)  //遍历，一一比较
	{
		if (cur->s.id == id)          //找到该节点
		{
			LPCYCLE_LINK_LIST temp;
			temp = cur;
			pre->next = temp->next;   //修改指向，就是删除了   
			free(temp);               //最后要释放节点
			--g_iCount;
			break;                    //记得退出，否则访问空指针会出错
		}
		cur = cur->next;
		pre = pre->next;
	}
}

//删除数据 （指定name）
void DeleteFromName(char* name)
{
	LPCYCLE_LINK_LIST cur = head.next;
	LPCYCLE_LINK_LIST pre = &head;
	for (int i = 0;i < g_iCount;i++)  //遍历，一一比较
	{
		if (strcmp(cur->s.name, name) == 0)          //找到该节点
		{
			LPCYCLE_LINK_LIST temp;
			temp = cur;
			pre->next = temp->next;   //修改指向，就是删除了   
			free(temp);               //最后要释放节点
			--g_iCount;
			break;
		}
		cur = cur->next;
		pre = pre->next;
	}
}

//查找数据 （指定id，返回name）
char* SearchFromId(int id)
{
	LPCYCLE_LINK_LIST cur = head.next;
	for (int i = 0;i < g_iCount;i++)  //遍历，一一比较
	{
		if (cur->s.id == id)          //找到该节点
			return cur->s.name;
		cur = cur->next;
	}
	return NULL;
}

//查找数据 （指定name，返回id）
int SearchFromName(char* name)
{
	LPCYCLE_LINK_LIST cur = head.next;
	for (int i = 0;i < g_iCount;i++)  //遍历，一一比较
	{
		if (strcmp(cur->s.name, name) == 0)          //找到该节点
			return cur->s.id;
		cur = cur->next;
	}
	return -1;
}

//根据local查找指定节点，返回该节点
LPCYCLE_LINK_LIST GetNode(int local)
{
	if (local < g_iCount && local >= 0)
	{
		LPCYCLE_LINK_LIST cur = head.next;
		for (int i = 0;i < local;i++)  //遍历
			cur = cur->next;
		return cur;
	}
	else
		return NULL;
}

//逆转表(方法1，牺牲时间换取空间，速度很慢，时间复杂度为0(n^2))
void Reverse1()
{
	STUDENT temp;
	for (int i = 0;i < g_iCount / 2;i++)
	{
		LPCYCLE_LINK_LIST node1 = GetNode(i);                //获取节点
		LPCYCLE_LINK_LIST node2 = GetNode(g_iCount - i - 1);
		temp = node1->s;                              //交换数据项，千万别交换节点
		node1->s = node2->s;
		node2->s = temp;
	}
}

//逆转表(方法2，牺牲空间换取时间，速度很快，时间复杂度为0(n))
void Reverse2()
{
	CYCLE_LINK_LIST tHead = { 0 };     //临时头节点
	tHead.next = &tHead;
	LPCYCLE_LINK_LIST cur = head.next;
	for (int i = 0;i < g_iCount;i++)
	{
		LPCYCLE_LINK_LIST l = (LPCYCLE_LINK_LIST)malloc(sizeof(_link_list));
		memcpy(l, cur, sizeof(_link_list));   //这里一定要复制，而不是赋值=，如果是=，则两指针指向的是同一个节点
		l->next = tHead.next;        //这里采用头添加，结束后，就会自动逆转过来了         
		tHead.next = l;
		cur = cur->next;
	}
	cur = head.next;
	LPCYCLE_LINK_LIST temp = tHead.next;
	while ((cur->next) != &tHead)
	{
		memcpy(cur, temp, sizeof(_link_list));   //因为头添加已经是逆转了，所以直接复制到原来的链表中即可
		temp = temp->next;
		cur = cur->next;
	}    
	cur->next = &head;//最后让尾节点指向头节点,若没有这一句，则cur->next指向的是tHead
}

//修改数据
void Rewrite(int local, STUDENT s)
{
	if (local < g_iCount && local >= 0)   //当要删除的位置合理的情况下才能删除
	{
		LPCYCLE_LINK_LIST cur = head.next;
		for (int i = 0;i < local;i++)    //遍历到指定位置
			cur = cur->next;
		cur->s = s;                       //修改
	}
}

//清空链表
void ClearLinkList()
{
	LPCYCLE_LINK_LIST cur = head.next;
	for (int i = 0;i < g_iCount;i++)
	{
		LPCYCLE_LINK_LIST temp;
		temp = cur;         //保存前一个节点的地址 
		cur = cur->next;    //向前查找下一个节点
		free(temp);         //释放前一个节点的空间
	}
	head.next = &head;      //最后让头节点指向头节点
}

//遍历数据
void Print()
{
	LPCYCLE_LINK_LIST cur = &head;
	while ((cur = cur->next) != &head)
		printf("%d - %s\n", cur->s.id, cur->s.name);
}
